llama_stack/apis/credentials/credentials.py

```diff
@@ -0,0 +1,86 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+# All rights reserved.
+#
+# This source code is licensed under the terms described in the LICENSE file in
+# the root directory of this source tree.
+
+from datetime import datetime
+from enum import Enum
+from typing import Protocol, runtime_checkable
+
+from pydantic import BaseModel, Field
+
+from llama_stack.schema_utils import json_schema_type, webmethod
+
+
+class CredentialTokenType(str, Enum):
+    """The type of credential token.
+
+    :cvar oauth2: OAuth2 token
+    :cvar api_key: API key
+
+    """
+
+    oauth2 = "oauth2"
+    api_key = "api_key"
+
+
+@json_schema_type
+class ProviderCredential(BaseModel):
+    credential_id: str
+    provider_id: str
+    token_type: CredentialTokenType
+    token: str
+    expires_at: datetime = Field(description="The time at which the credential expires. In UTC.")
+
+
+@runtime_checkable
+class Credentials(Protocol):
+    """
+    Create, update and delete ephemeral provider-specific credentials.
+
+    Each provider may need optional authentication. This might be a persistent API key, or
+    a short-lived OAuth2 token. There is a single credential for each provider instance.
+
+    Credentials are ephemeral -- they may be purged after the specified TTL.
+
+    Credentials are associated with the same ABAC access attributes and permissions as other
+    resources in the system.
```

Comment:
There is no explicit association between a user and a credential. I assume that would be implied by the principal of the token used to authorize the request that creates the credential? What would be the use case for some other user to update the token belonging to a different user? (I assume that update, in contrast to create, would not alter the association of the credential to a particular user).

Reply:
@grs Consider the first use case I specified above -- you may want to share the same Fireworks API key across your team.

Reply:
How is the association between a user (a principal) and a credential established? I.e. for an agent created by user 'foo', how is the correct credential for an mcp server with id 'bar' determined? Does it just check whether any credential with that provider id has access attributes allowing it to be used (and picks the first if there are more than one)?

Reply:
yeah it is stored against the logged-in-user. You will see that in the next PR.